### PR TITLE
Clean harness option loading

### PR DESCRIPTION
**Pull request by mhegazy**
following up on the conversation in https://github.com/Microsoft/TypeScript/pull/4371#discussion_r37496529, this consolidates all compiler option parsing in one place in the harness.

### CODE REVIEW

File: src/harness/harness.ts [hunk 1]
```diff
@@ -910,233 +910,127 @@ module Harness {
...
-                let program = ts.createProgram(programFiles, options, compilerHost);
-
-                let emitResult = program.emit();
-
-                let errors = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
-                this.lastErrors = errors;
-
-                let result = new CompilerResult(fileOutputs, errors, program, ts.sys.getCurrentDirectory(), emitResult.sourceMaps);
-                onComplete(result, program);
+        interface HarnesOptions {
```
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: `HarnessOptions`

File: src/harness/harness.ts [hunk 2]
```diff
@@ -1144,49 +1038,176 @@ module Harness {
...
+                            throw new Error("Unsupported compiler setting " + value);
                     }
                 }
             }
+        }
+  
+        export class HarnessCompiler {
+            private inputFiles: { unitName: string; content: string }[] = [];
+            private compileOptions: ts.CompilerOptions;
+            private settings: Harness.TestCaseParser.CompilerSettings = {};
```
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: Maybe to keep these in sync with `reset`, the constructor should call `reset`

File: src/harness/harness.ts [hunk 3]
```diff
@@ -1144,49 +1038,176 @@ module Harness {
...
+            public reportCompilationErrors() {
+                return this.lastErrors;
+            }
+
+            public setCompilerSettings(tcSettings: Harness.TestCaseParser.CompilerSettings) {
+                this.settings = tcSettings;
+            }
+
+            public addInputFiles(files: { unitName: string; content: string }[]) {
+                files.forEach(file => this.addInputFile(file));
```
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: `forEach` or `.forEach`? I always forget what the rule is for the harness

Comment (mhegazy) [inline src/harness/harness.ts]: this did not change here.. that was just moving code around

**Comment by mhegazy** [comment]
So i have moved to just use the commandlineParser defintions instead of parsing each string into an enum. this way any compiler option added is a valid harness option.

**Comment by sheetalkamat** [comment]
:+1:

File: src/harness/harness.ts [hunk 4]
```diff
@@ -949,16 +949,76 @@ module Harness {
...
+                                break;
+                            // If not a primitive, the possible types are specified in what is effectively a map of options.
+                            default:
+                                let map = <ts.Map<number>>option.type;
+                                let key = (value).toLowerCase();
+                                if (ts.hasProperty(map, key)) {
+                                    options[option.name] = map[key];
+                                }
+                                else {
+                                    throw new Error(`Unkown value '${value}' for compiler option '${name}'.`);
```
Comment (vladima) [inline src/harness/harness.ts]: 'Unknown'

File: src/harness/fourslash.ts [hunk 5]
```diff
@@ -117,89 +117,17 @@ module FourSlash {
...
-        jsx: "jsx",
     };
 
     // List of allowed metadata names
     let fileMetadataNames = [metadataOptionNames.fileName, metadataOptionNames.emitThisFile, metadataOptionNames.resolveReference];
-    let globalMetadataNames = [metadataOptionNames.allowNonTsExtensions, metadataOptionNames.baselineFile, metadataOptionNames.declaration,
-        metadataOptionNames.mapRoot, metadataOptionNames.module, metadataOptionNames.out,metadataOptionNames.outFile,
-        metadataOptionNames.outDir, metadataOptionNames.sourceMap, metadataOptionNames.sourceRoot, metadataOptionNames.jsx];
 
     function convertGlobalOptionsToCompilerOptions(globalOptions: { [idx: string]: string }): ts.CompilerOptions {
```
Comment (vladima) [inline src/harness/fourslash.ts]: should type of `globalOptions` here also become `Harness.TestCaseParser.CompilerSettings` similar to the change in lines 32/38 ?

File: src/harness/harness.ts [hunk 6]
```diff
@@ -949,16 +949,76 @@ module Harness {
             };
         }
 
+        interface HarnesOptions {
```
Comment (vladima) [inline src/harness/harness.ts]: `HarnessOptions`

File: src/harness/harness.ts [hunk 7]
```diff
@@ -949,16 +949,76 @@ module Harness {
...
+                            case "boolean":
+                                options[option.name] = value.toLowerCase() === "true";
+                                break;
+                            case "string":
+                                options[option.name] = value;
+                                break;
+                            // If not a primitive, the possible types are specified in what is effectively a map of options.
+                            default:
+                                let map = <ts.Map<number>>option.type;
+                                let key = (value).toLowerCase();
```
Comment (vladima) [inline src/harness/harness.ts]: value is guaranteed to be not-null\undefined?

File: src/harness/harness.ts (hunk 4 above)
Comment (mhegazy) [inline src/harness/harness.ts]: done.

File: src/harness/fourslash.ts (hunk 5 above)
Comment (mhegazy) [inline src/harness/fourslash.ts]: done.

File: src/harness/harness.ts (hunk 6 above)
Comment (mhegazy) [inline src/harness/harness.ts]: done.

**Comment by vladima** [comment]
:+1:

File: src/harness/harness.ts (hunk 7 above)
Comment (mhegazy) [inline src/harness/harness.ts]: added a check.

**Pull request merged**